src/test/ui/borrowck/regions-escape-bound-fn.stderr

```diff
@@ -0,0 +1,12 @@
+error: borrowed data cannot be stored outside of its closure
+  --> $DIR/regions-escape-bound-fn.rs:18:27
+   |
+18 |     with_int(|y| x = Some(y));
+   |              ---     -----^-
+   |              |       |    |
+   |              |       |    cannot be stored outside of its closure
+   |              |       cannot infer an appropriate lifetime
+   |              borrowed data cannot outlive this closure
```

Comment:
@nikomatsakis I'm not entirely happy with the output on this case :-/

This triggers [the case I originally neglected](https://github.com/rust-lang/rust/pull/47144/files#diff-e5107c6d2a404ecd34f782e9e9e8b20f) and was trying to fix...

Should we keep this PR on hold until I have time to look at how to differentiate between these two cases for a more appropriate output? In my mind, this case should completely ignore the lifetime issue for `Some(_)` and point at the destination binding, like in the output of the original case for this PR:

```

error: borrowed data cannot be stored outside of its closure
  --> $DIR/regions-escape-bound-fn.rs:18:27
   |
17 |     let mut x: Option<&isize> = None;
   |            - borrowed data cannot be stored into here...
18 |     with_int(|y| x = Some(y));
   |              ---          ^ cannot be stored outside of its closure
   |              |
   |              ...because it cannot outlive this closure

```

It should be possible to do so as [the current diagnostic](https://play.rust-lang.org/?gist=844fbf39824c9fd8170b6c1eda08121d&version=nightly) does point at `x`'s declaration:

```
error[E0495]: cannot infer an appropriate lifetime for automatic coercion due to conflicting requirements
 --> src/main.rs:8:22
  |
8 |     with_int(|y| x = Some(y));
  |                      ^^^^^^^
  |
note: first, the lifetime cannot outlive the anonymous lifetime #2 defined on the body at 8:14...
 --> src/main.rs:8:14
  |
8 |     with_int(|y| x = Some(y));
  |              ^^^^^^^^^^^^^^^
note: ...so that reference does not outlive borrowed content
 --> src/main.rs:8:27
  |
8 |     with_int(|y| x = Some(y));
  |                           ^
note: but, the lifetime must be valid for the block suffix following statement 0 at 7:5...
 --> src/main.rs:7:5
  |
7 | /     let mut x: Option<&isize> = None;
8 | |     with_int(|y| x = Some(y));
9 | |     //~^ ERROR borrowed data cannot be stored outside of its closure
10| | }
  | |_^
note: ...so that variable is valid at time of its declaration
 --> src/main.rs:7:9
  |
7 |     let mut x: Option<&isize> = None;
  |         ^^^^^
```

Also, should I create a new error code for each of these cases, one shared between all of them or reuse the existing one?

Reply:
I think this output is way better than the original output. The main change I would do is not to print the "cannot infer an appropriate lifetime" label, which seems to add zero value. Saying that the data cannot escape the closure and indicating the closure is pretty awesome.